src/main/java/com/attacktimer/VariableSpeed/VariableSpeed.java

```diff
@@ -0,0 +1,78 @@
+package com.attacktimer.VariableSpeed;
+
+/*
+ * Copyright (c) 2024, Lexer747 <https://github.com/Lexer747>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import com.attacktimer.AttackProcedure;
+import com.attacktimer.AnimationData;
+import net.runelite.api.Client;
+import net.runelite.api.events.GameTick;
+
+public class VariableSpeed {
+
+    /**
+     * computeSpeed will forward the client, animation data and current weapon speed to all the known classes
+     * which can affect the base speed of a weapon. See implementations of IVariableSpeed.
+     * @param client
+     * @param curAnimation
+     * @param atkType
+     * @param baseSpeed
+     * @return
+     */
+    public static int computeSpeed(final Client client, final AnimationData curAnimation, final AttackProcedure atkProcedure, final int baseSpeed)
+    {
+        int newSpeed = baseSpeed;
+        for (IVariableSpeed i : toApply)
+        {
+            newSpeed = i.apply(client, curAnimation, atkProcedure, baseSpeed, newSpeed);
+        }
+        return newSpeed;
+    }
+
+    public static void onGameTick(Client client, GameTick tick)
+    {
+        for (IVariableSpeed i : toApply)
+        {
+            i.onGameTick(client, tick);
+        }
+    }
+
+    private static final IVariableSpeed[] toApply = {
+        // Order matters, apply leagues first, then any incremental modifications like rapid, or set effects.
+        // Then overriding speeds last, which set a speed.
+        new Leagues(),

```

Comment:
I think rapid should be applied before leagues modifiers.

Reply:
It's not actually clear to me which is correct, and mathematically because of the rounding it's always the same effective computation no matter which way round you do the maths. I think until there's a concrete weapon which proves the maths is done a certain way we can leave the order arbitrary.

E.g. Darts:

1. 3 ticks, div 2 rounded up is 2 ticks, minus 1 tick for rapid is 1 tick overall.
2. 3 ticks, minus 1 tick for rapid, div 2 rounded up is 1 tick overall.